File: Data Structures & Algorithms/implement-prefix-tree/submission-1.java
class TrieNode{
    private HashMap<Character, TrieNode> children;
    private boolean endOfWord;

    public TrieNode(){
        this.children = new HashMap<>();
        this.endOfWord = false; 
    }
}

class PrefixTree {

    private TrieNode root;

    public PrefixTree() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode curr = this.root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
            if(!curr.children.containsKey(c)){
                curr.children.put(c, new TrieNode());
            }
            curr = curr.children.get(c);
        }
        curr.endOfWord = true;
    }

    public boolean search(String word) {
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
            if(!curr.children.containsKey(c)){
                return false;
            }
            curr = curr.children.get(c);
        }
        return curr.endOfWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for(int i = 0; i < prefix.length(); i++){
            char c = prefix.charAt(i);
            if(!curr.children.containsKey(c)){
                return false;
            }
            curr = curr.children.get(c);
        }
        return true;
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-1.py
class Node:
    def __init__(self):
        self.children = {} 
        self.endOfWord = False

class WordDictionary:

    def __init__(self):
        self.root = Node() # [{d: [{a: [{y: [{}, True]}, False]}, False]}, False]

    def addWord(self, word: str) -> None: # day
        curr = self.root
        for c in word: # a
            if c not in curr.children:
                curr.children[c] = Node()
            curr = curr.children[c]
        curr.endOfWord = True
        
    def search(self, word: str) -> bool:
        # word = day, pos = 0
        def dfs(pos, node):
            curr = node # curr = {root}

            for i in range(pos, len(word)):
                c = word[i]

                if c == ".":
                    for child in curr.children.values():
                        if dfs(i + 1, child):
                            return True
                    return False

                else:
                    if c not in curr.children:
                        return False
                    curr = curr.children[c]

            return curr.endOfWord

        return dfs(0, self.root)
        
